recognition/include/pcl/recognition/impl/linemod/line_rgbd.hpp

```diff
@@ -150,72 +150,59 @@ LineRGBD<PointXYZT, PointRGBT>::loadTemplates (const std::string &file_name, con
-    float max_z = -std::numeric_limits<float>::max ();
-    std::size_t counter = 0;
-    for (std::size_t j = 0; j < template_point_cloud.size (); ++j)
-    {
-      const PointXYZRGBA & p = template_point_cloud[j];
-
-      if (!std::isfinite (p.x) || !std::isfinite (p.y) || !std::isfinite (p.z))
-        continue;
-
-      min_x = std::min (min_x, p.x);
-      min_y = std::min (min_y, p.y);
-      min_z = std::min (min_z, p.z);
-      max_x = std::max (max_x, p.x);
-      max_y = std::max (max_y, p.y);
-      max_z = std::max (max_z, p.z);
+    bounding_boxes_[i] = computeBoundingBoxAndCenterTemplatePointCloud (template_point_clouds_[i]);
+  }
 
-      center_x += p.x;
-      center_y += p.y;
-      center_z += p.z;
+  return (true);
+}
 
-      ++counter;
-    }
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointXYZT, typename PointRGBT> BoundingBoxXYZ
+pcl::LineRGBD<PointXYZT, PointRGBT>::computeBoundingBoxAndCenterTemplatePointCloud (PointCloud<PointXYZRGBA> & template_point_cloud)
+{
+  BoundingBoxXYZ bb;
+  bb.x = bb.y = bb.z = std::numeric_limits<float>::max ();
+
+  Eigen::Vector4f geometric_center = Eigen::Vector4f::Zero ();
+  Eigen::Vector4f min_pos, max_pos;
+  min_pos.fill (std::numeric_limits<float>::max ());
+  max_pos.fill (std::numeric_limits<float>::lowest ());
+  std::size_t counter = 0;
+  for (const PointXYZRGBA & p : template_point_cloud.points)
+  {
+    if (!isXYZFinite(p))
+      continue;
 
-    center_x /= static_cast<float> (counter);
-    center_y /= static_cast<float> (counter);
-    center_z /= static_cast<float> (counter);
+    min_pos = min_pos.cwiseMin (p.getVector4fMap ());
+    max_pos = max_pos.cwiseMax (p.getVector4fMap ());
 
-    bb.width  = max_x - min_x;
-    bb.height = max_y - min_y;
-    bb.depth  = max_z - min_z;
+    geometric_center += p.getVector4fMap ();
 
-    bb.x = (min_x + bb.width / 2.0f) - center_x - bb.width / 2.0f;
-    bb.y = (min_y + bb.height / 2.0f) - center_y - bb.height / 2.0f;
-    bb.z = (min_z + bb.depth / 2.0f) - center_z - bb.depth / 2.0f;
+    ++counter;
+  }
+  geometric_center /= static_cast<float> (counter);
 
-    for (std::size_t j = 0; j < template_point_cloud.size (); ++j)
-    {
-      PointXYZRGBA p = template_point_cloud[j];
+  Eigen::Vector4f bb_dim = max_pos - min_pos;
+  bb.width  = bb_dim[0];
+  bb.height = bb_dim[1];
+  bb.depth  = bb_dim[2];
 
-      if (!std::isfinite (p.x) || !std::isfinite (p.y) || !std::isfinite (p.z))
-        continue;
+  Eigen::Vector4f diff_pos = min_pos - geometric_center;
+  bb.x = diff_pos[0];
+  bb.y = diff_pos[1];
+  bb.z = diff_pos[2];
 
-      p.x -= center_x;
-      p.y -= center_y;
-      p.z -= center_z;
+  for (PointXYZRGBA & p : template_point_cloud.points)
+  {
+    if (!isXYZFinite(p))
+      continue;
 
-      template_point_cloud[j] = p;
-    }
+    p.getVector4fMap () -= geometric_center;
```

Comment:
No need to check if the point is finite. if it isn't the subtraction operation will generate more nans on the same points.